### PR TITLE
Use AsyncDataCache and read-ahead with Hive

### PR DESCRIPTION
**Pull request by oerling**
Uses AsyncDataCache and ScanTracker with Hive if the MappedMemory
given to the reader is a AsyncDataCache. If this is a regular
MappedMemory, uses BufferedInput with the given DataCache, if any.

Sets most tests to use AsyncDataCache as MappedMemory also if no
affinity.

Tests TableScan with both AsyncDataCache and regular MappedMemory +
SimpleLruDataCache.

### CODE REVIEW

File: velox/connectors/Connector.cpp [hunk 1]
```diff
@@ -73,4 +73,32 @@ std::shared_ptr<Connector> getConnector(const std::string& connectorId) {
...
+    const std::string& scanId) {
+  return trackers_.withWLock([&](auto& trackers) -> auto {
+    auto it = trackers.find(scanId);
+    if (it == trackers.end()) {
+      auto newTracker =
+          std::make_shared<cache::ScanTracker>(scanId, unregisterTracker);
+      trackers[newTracker->id()] = newTracker;
+      return newTracker;
+    }
+    std::shared_ptr tracker = it->second.lock();
```
Comment (mbasmanova) [inline velox/connectors/Connector.cpp:95]: std::shared_ptr -> std::shared_ptr<cache::ScanTracker>?

File: velox/connectors/Connector.h [hunk 2]
```diff
@@ -35,6 +38,10 @@ namespace facebook::velox::connector {
 struct ConnectorSplit {
   const std::string connectorId;
 
+  // true if the Task processing this has aborted. Allows aborting
+  // async prefetch for the split.
+  bool cancelled{};
```
Comment (mbasmanova) [inline velox/connectors/Connector.h:43]: {} -> {false} ?

File: velox/connectors/Connector.h [hunk 3]
```diff
@@ -133,10 +140,14 @@ class ConnectorQueryCtx {
   ConnectorQueryCtx(
       memory::MemoryPool* pool,
       Config* config,
-      ExpressionEvaluator* expressionEvaluator)
+      ExpressionEvaluator* expressionEvaluator,
+      memory::MappedMemory* mappedMemory,
+      std::optional<std::string> scanId = std::nullopt)
```
Comment (mbasmanova) [inline velox/connectors/Connector.h:145]: Why scanId is optional? Can't we always populate it?

File: velox/connectors/Connector.h [hunk 4]
```diff
@@ -150,10 +161,25 @@ class ConnectorQueryCtx {
     return expressionEvaluator_;
   }
 
+  memory::MappedMemory* mappedMemory() const {
```
Comment (mbasmanova) [inline velox/connectors/Connector.h:164]: Would you document what is this to be used for?

File: velox/connectors/hive/HiveConnector.cpp [hunk 5]
```diff
@@ -445,17 +504,34 @@ void HiveDataSource::setNullConstantValue(
...
           std::make_unique<SimpleLRUCache<std::string, FileHandle>>(
               FLAGS_file_handle_cache_mb << 20),
-          std::make_unique<FileHandleGenerator>(properties)) {}
+          std::make_unique<FileHandleGenerator>()),
+      executor_(executor) {}
 
 VELOX_REGISTER_CONNECTOR_FACTORY(std::make_shared<HiveConnectorFactory>())
-
 } // namespace facebook::velox::connector::hive
+  // namespace facebook::velox::connector::hive
```
Comment (mbasmanova) [inline velox/connectors/hive/HiveConnector.cpp:537]: duplicate coment; remove

File: velox/exec/TableWriter.cpp [hunk 6]
```diff
@@ -41,7 +41,8 @@ TableWriter::TableWriter(
   // partition
   const auto& connectorId = tableWriteNode->insertTableHandle()->connectorId();
   connector_ = connector::getConnector(connectorId);
-  connectorQueryCtx_ = driverCtx_->createConnectorQueryCtx(connectorId);
+  connectorQueryCtx_ =
+      driverCtx_->createConnectorQueryCtx(connectorId, "TableWriter");
```
Comment (mbasmanova) [inline velox/exec/TableWriter.cpp:45]: This is strange. The second parameter is supposed to be the ID of the TableScan node, but it is "TableWriter".

**Review by mbasmanova — APPROVED** [review]
Looks good overall, but please fix minor comments below before landing.

**Comment by oerling** [comment]
Thank you. I made the edits.

**Pull request closed without merging**